File: cses/01_introductory/3311_grid-coloring-i/solution.py
import sys


def main():
    # Đọc toàn bộ input; n, m rồi n hàng ký tự (mỗi hàng là bytes chỉ gồm A,B,C,D)
    data = sys.stdin.buffer.read().split()
    n = int(data[0])
    m = int(data[1])
    rows = data[2:2 + n]

    colors = b'ABCD'
    out = []
    prev_row = None  # hàng trước đã tô (bytearray các mã ASCII)

    # Duyệt row-major, tô tham lam từng ô
    for i in range(n):
        orig = rows[i]
        cur = bytearray(m)
        for j in range(m):
            # Ba ràng buộc: khác màu gốc, khác ô trên, khác ô trái
            forbid_o = orig[j]
            up = prev_row[j] if prev_row is not None else 0
            left = cur[j - 1] if j > 0 else 0
            # Chọn màu đầu tiên hợp lệ (luôn tồn tại vì có 4 màu, tối đa 3 bị cấm)
            for c in colors:
                if c != forbid_o and c != up and c != left:
                    cur[j] = c
                    break
        out.append(cur.decode())
        prev_row = cur

    sys.stdout.write('\n'.join(out))
    sys.stdout.write('\n')


main()

File: cses/01_introductory/3311_grid-coloring-i/solution.cpp
#include <array>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    // Đọc kích thước lưới và lưới ký tự gốc
    int n, m;
    cin >> n >> m;
    vector<string> original(n);
    for (string& row : original) {
        cin >> row;
    }

    const array<char, 4> colors = {'A', 'B', 'C', 'D'};
    vector<string> answer(n, string(m, '?'));

    // Duyệt row-major, tô tham lam từng ô
    for (int row = 0; row < n; ++row) {
        for (int col = 0; col < m; ++col) {
            // Chọn màu đầu tiên khác màu gốc, khác ô trên và khác ô trái
            for (const char color : colors) {
                const bool sameAsOriginal = color == original[row][col];
                const bool sameAsAbove = row > 0 && color == answer[row - 1][col];
                const bool sameAsLeft = col > 0 && color == answer[row][col - 1];
                if (!sameAsOriginal && !sameAsAbove && !sameAsLeft) {
                    answer[row][col] = color;
                    break;
                }
            }
        }
    }

    // Luôn tồn tại lời giải nên chỉ cần in lưới kết quả
    for (const string& row : answer) {
        cout << row << '\n';
    }

    return 0;
}
